Consistent statusline spacing

hamrik opened the pull request:
Statusline elements used to have inconsistent spacing due to each element having to pad itself without knowledge of surrounding elements.

<img alt="Inconsistent spacing" src="https://github.com/helix-editor/helix/assets/18398499/8cca6179-4593-47e6-8a7f-acfe6c7e89c8">

 - `mode`, `selections`, `file-name` and `position` are padded, the rest are not
 - `file-modification-indicator` reserves space for itself, `spinner` and `diagnostics` do not

This PR fixes these spacing inconsistencies.

<img alt="Consistent spacing" src="https://github.com/helix-editor/helix/assets/18398499/f06f7848-b500-4901-b5de-a5ed0328d116">

 - No elements pad themselves
 - No elements reserve empty space
 - The statusline pads everything
 - The `spacer` element has been repurposed to join multiple spacers, e.g. to pad the ends of the statusline

The render functions of statusline elements previously used an ad-hoc `write` callback to blit `Spans` directly to the `Surface` of the statusline. This callback has been removed. Now each render function has to return the `Spans` it wants to display, which the renderer of the statusline collects, intersperses with spaces, then blits to the surface.

gabydd commented:
At first look one thing that this changes that we probably shouldn't is it removes padding around the edges in the default config which imo we should keep especially looks bad when ``color-modes`` is false. This brings up another question if we are going for consistency do we really want ``mode``s size to be dependant on ``coior-modes``? I haven't looked much at the implementation but I think improving consistency is generally good as long as we are not breaking peoples configs.

David-Else commented:
@gabydd Regarding `color -modes`: https://github.com/helix-editor/helix/discussions/7289#discussioncomment-6124763

gabydd commented:
Ah right, I only noticed this because I don't use color-modes. Removing the option is probably a good idea in the long run though.

the-mikedavis commented on helix-tui/src/text.rs:
```diff
@@ -229,6 +229,70 @@ impl<'a> Spans<'a> {
+
+    /// Constructs a single, styled `Span` with the given `string` and `style`
+    /// and wraps it in a `Spans`.
+    pub fn single<T>(string: T, style: Style) -> Spans<'a>
+    where
+        T: Into<Cow<'a, str>>,
+    {
+        Spans(vec![Span::styled(string, style)])
+    }
```
This already exists as a `From`: https://github.com/helix-editor/helix/blob/2a11fb485d33eb33a7743899e9a86f745f6d5e02/helix-tui/src/text.rs#L258-L262

Places where this function is used can either use `Span::styled(string, style).into()` or explicitly call `Spans::from` on that Span.

hamrik replied:
Didn't know, will change. Thanks!

hamrik replied:
Fixed in 560fd9aa

the-mikedavis commented on helix-tui/src/text.rs:
````diff
@@ -229,6 +229,70 @@ impl<'a> Spans<'a> {
+    /// spans.push("Hello!", Style::default());
+    /// assert_eq!("Hello!", spans.0[0].content);
+    /// ```
+    pub fn push<T>(&mut self, string: T, style: Style)
````
This should take a `Span<'a>` instead

hamrik replied:
It makes sense, but that will bloat the current usages a bit. I guess I could always add a helper function there instead. Will change and see if a utility function becomes necessary in `statusline.rs`.

hamrik replied:
Fixed in 8e8d860b. Boilerplate is minimal, no extra helper added.

the-mikedavis commented on helix-tui/src/text.rs:
````diff
@@ -229,6 +229,70 @@ impl<'a> Spans<'a> {
+    /// );
+    /// assert_eq!("Hello World!", String::from(spans));
+    /// ```
+    pub fn join<S>(mut spanses: S, sep: &Spans<'a>) -> Spans<'a>
````
```suggestion
    pub fn join<S>(mut spans_iter: S, sep: &Spans<'a>) -> Spans<'a>
```

the-mikedavis replied:
Does `sep` need to be a `Spans` here? I think it could just be a `Span`.

hamrik replied:
Using spans allows for fancier separators, for example one with multiple colors.

I think I will make two more versions of `join`, one taking a `Span` as you suggest, and one taking an interpolator function which takes two neighboring `Spans` and generates a separator for them. This could be useful for powerline-style color transitions.

hamrik replied:
Variable naming fixed in 8e8d860b

hamrik replied:
Removed all custom Spans functions so this issue is now resolved.

the-mikedavis commented on helix-tui/src/text.rs:
```diff
@@ -229,6 +229,70 @@ impl<'a> Spans<'a> {
+        S: Iterator<Item = Spans<'a>>,
+    {
+        let mut spans: Vec<Span> = Vec::new();
+        if let Some(head) = spanses.next() {
+            spans.extend(head.0);
+        }
+        spanses.for_each(|s| {
+            spans.extend(sep.0.clone());
+            spans.extend(s.0);
+        });
```
It's a style nit but `for_each` is generally considered less idiomatic in Rust than regular `for` or `while` loops

```rust
while let Some(spans) = span_iter.next() {
    if !new_spans.is_empty() {
        new_spans.push(sep.clone());
    }
    new_spans.extend(spans);
}
```

(see the note in the rust docs: https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.for_each)

hamrik replied:
Will change, thanks!

hamrik replied:
Fixed in 8e8d860b

the-mikedavis commented on book/src/configuration.md:
```diff
@@ -116,7 +118,7 @@ The following statusline elements can be configured:
 | `position-percentage` | The cursor position as a percentage of the total number of lines |
 | `separator` | The string defined in `editor.statusline.separator` (defaults to `"│"`) |
-| `spacer` | Inserts a space between elements (multiple/contiguous spacers may be specified) |
+| `spacer` | An empty string, used to join multiple spacers |
```
```suggestion
| `spacer` | A string used as space to separate all statusline elements, defined in `editor.statusline.spacer` (defaults to `" "`) |
```

hamrik replied:
The option has been removed, so this is no longer an issue.

pascalkuthe replied:
I think I prefer the original docs here. The new docs aren't quite accurate and since the behavior hasn't been changed I see no reason to change the docs either

gabydd replied:
Yeah even if this is what it technically does code wise, the docs are user facing and should explain what is actually important to user which is it adds another space

hamrik replied:
Ended up leaving the original description.

pascalkuthe commented:
I am not sure why we need the `editor.statusline.spacer` config. Why would this be ever set to anything but space?

hamrik commented:
Sorry for disappearing, had a busy couple of weeks. Checking the responses right now.

hamrik commented:
> At first look one thing that this changes that we probably shouldn't is it removes padding around the edges in the default config which imo we should keep especially looks bad when `color-modes` is false. This brings up another question if we are going for consistency do we really want `mode`s size to be dependant on `coior-modes`? I haven't looked much at the implementation but I think improving consistency is generally good as long as we are not breaking peoples configs.

I originally didn't have padding in `color-modes` either, but then a color cuts right off at the end of the mode name, which didn't look right. I agree that this makes `modes` a bit inconsistent and will have to think of something to address this. 

Padding the statusbar edges could be done by changing the default config to include `spacer`-s on the front or back, or alternatively I could for a unit of padding into the statusline. With the latter option `color-modes` won't look right and the user wouldn't be able to fix it by removing the `spacer`, so I vote for chaging the default config. Those who have customized statuslines shouldn't have a hard time adding a spacer if they don't use `color-modes`.

hamrik commented on helix-term/src/ui/statusline.rs:
```diff
@@ -164,162 +155,130 @@ where
-                    .get(srv.id())
-                    .and_then(|spinner| spinner.frame())
-            })
-            // Even if there's no spinner; reserve its space to avoid elements frequently shifting.
-            .unwrap_or(" ")
```
But if the behavior is kept then there will be a large empty space between the elements surrounding the spinner when the spinner is not in use.

I guess either situation could be solved by simply placing the spinner as the last item of the left side or the first item of the right side, blending it in with the surrounding spacing. Or we could change the spinner idle frame to always display a faded out placeholder when not in use, e.g. `⣿`.

twharmon replied:
Items already jump around for other reasons, right? Almost very single element that can go in the status line can cause shifts (diagnostics, selections, position, etc.).

I simply put the elements that cause the most shift or most frequent shift (diagnostics) at the end of the list.

hamrik replied:
No feedback arrived for a while on this, and since I agree with @twharmon I will mark this as resolved for now. 

As I said, if the spinner absolutely _must_ reserve space we should use a visible idle character, but it would be more elegant to simply place the spinner on the inner edge of the side section where it won't push other elements around.

hamrik commented on helix-tui/src/text.rs:
```diff
@@ -229,6 +229,70 @@ impl<'a> Spans<'a> {
+        self.0.push(Span::styled(string, style));
+    }
+
+    /// Creates a new, empty `Spans`.
+    pub fn new() -> Spans<'a> {
+        Spans(vec![])
+    }
```
Didn't know, will change. Thanks!

hamrik replied:
Fixed in 560fd9aa

hamrik commented:
> I am not sure if we need the `editor.statusline.spacer` config. Why would this be ever set to anything but space?

A couple examples:
  - Two or more spacers for wider spacing without spamming `spacer`: `file.txt   1:1   [+]`
  - Powerline-style (colorless) separators: `file.txt 〉1:1 〉[+]`
    - Altough to make this last one flippable we would have to add separate spacer configs for left, center and right.

pascalkuthe commented:
> 

Generally I would prefer explicit spacers since it allows different spacing between different elements without adding a ton new config options. I don't think that the spacing inside of elements needs to be configurable. We are shooting for a mininal config file.

pascalkuthe commented:
> You are free to leave `editor.statusline.spacer` on the default setting and use spacers. It don't feel it's out of place next to the configurable separator and configurable mode names.

we are very conservative with regards to adding new config options and this doens' cross the bar for me. The rest of the PR is probably a nice improvement but I don't see us adding this config option

hamrik commented:
Alright, I've removed the option to keep things moving along. At least that resolved the documentation wording problem. We can revisit this option in a later PR.

hamrik commented:
I believe the only two remaining issues are the spacing of the mode indicator and the spinner. 

I don't like the idea of hardcoding spacers on the edges, in my opinion modifying the default config to include spacers would be more flexible.

But regardless of how the padding is handled, it would look weird the color modes option turned on.

If we go with the hardcoded route, we could omit adding spacers where the mode indicator is the first/last item and color modes are enabled. But this feels like a hack.

If we go with the modified default config, we could add/omit the spacer depending on whether we enable color modes by default, then expect people to add/remove the spacer manually if they change the color modes option.

On a similar note, right now padding is automatically added to the mode indicator when color-modes are on, but that could also be removed if we expect people to modify the mode names to include spaces when they toggle color modes.

Thoughts?

hamrik commented:
I have removed the auto-padding from the `mode` element, regardless of `color-modes`.

 - This makes the `mode` element consistent with the other elements.
 - If `color-modes` is off by default, people who turn it on can add padding via `editor.statusline.mode`.
 - If `color-modes` becomed default, the default values for mode names can also be tweaked to include spaces.

pascalkuthe commented:
Yeah I agree with that. I think hardcoded padding is never a good idea and changing the modenames is a pretty good workaround IMO

pascalkuthe reviewed (commented):
This PR is very hard to review right now because you changed the order of some.functions (and in general have included some unnecessary changes) sk the diff is pretty clobbered right now.

Could you undo that so it's easier for us to see what's actually changed

pascalkuthe commented:
Also I am not really sure if we really want the implhct spacer between each element. I actually lprefer having no implicit spacer at all and instead including the necessary specaers in the default config. This offers the most flexibility and is the easiet to implement

hamrik commented:
> I actually lprefer having no implicit spacer at all and instead including the necessary specaers in the default config.

Since some elements only appear occasionally (such as `file-modification-indicator`, `spinner`, etc) these would leave two spacers behind whenever they are hidden (assuming the config has explicit spacers between every element). Implementing collapsing spacers would be even more complicated IMO.

hamrik commented:
> This PR is very hard to review right now because you changed the order of some.functions (and in general have included some unnecessary changes) sk the diff is pretty clobbered right now.

I'll clean up the commit history on the weekend.

hamrik commented:
Okay, I've removed all additions to `Spans`, and consolidated the changes into two commits/"phases":

1) Refactor statusline elements to return `Spans` rather than using a callback. At this point the spacing behavior was identical to the previous implementation.
2) Removed all padding.

Right now there is no spacing interspersed as `Iterator::intersperse()` is in nightly for now. While this matches what @pascalkuthe requested I think spaces should be added because manually interspersed spacers don't collapse when an element hides itself, such as `file-modification-indicator`, leaving two spacers next to each other, which looks inconsistent.

Will address that later.

hamrik commented:
Added back automatic interspersing of spaces. Does not pad the sides of the panel, those should be done in config.

Uses a custom interspersing iterator until `instersperse()` lands in stable Rust, and I'm fairly certain that its implementation should be moved elsewhere, not yet sure where would be best.

As for the spinner, I think simply hiding it would be most consistent with other autohiding elements like `file-modification-indicator`. If the jumpyness is annoying it can be moved to a place where it does not push other elements around, such as the inner edge of the left and right sections.

hamrik commented:
The only thing remaining to tweak in my opinion is the default config. After that I'll reopen the PR.

On the left section:
- If `color-modes` becomes the default then the default mode strings should be padded with strings.
- If it does not than an extra spacer needs to be added to the front of the default element list.

On the right section:
- A spacer needs to be appended to the default element list.

hamrik commented:
Added spacers the default config to pad the statusline, with the assumtion that color modes will remain off by default for now.

People who want to enable it and have it look good will have to:
 1. Enable `color-modes`
 2. Set custom mode strings that include spaces, e.g. `normal = " NOR "`
 3. Set a custom `left` section that does not begin with `spacer`

Also fixed the spacer that added an extra space by accident since the cleanup.

Debating whether this should be documented. Also the intersperse implementation still needs a better place,
but I think it's OK to reopen the PR by this point.

hamrik commented:
I was thinking about `color-modes1 lately. Since the goal is to eventually remove the option altogether, maybe this would be a good time to set it to `true` by default and incorporate the rest of the tweaks mentioned above into the default config.

Then it would still look better for less effort for people who turn the feature off, then it does now when people turn it on without the tweaks.

hamrik commented:
Looks like `intersperse` won't make into the stdlib anytime soon so I introduced the `itertools` dependency to avoid reinventing the wheel.

If this is a problem I can revert the last commit, before which a hand-rolled solution was used add spacers between statusline elements.

the-mikedavis commented:
I would like to avoid itertools if possible - it's a fair amount of code to add for things that can usually be written inline. Here I think it's cleaner to use a helper function that returns Spans:

```rust
fn join_with_spaces<'a, I: Iterator<Item = Span<'a>>>(iter: I) -> Spans<'a> {
    let mut spans = Vec::new();
    for elem in iter {
        if !spans.is_empty() {
            spans.push(Span::raw(" "));
        }
        spans.push(elem);
    }
    spans.into()
}

// ..

context.parts.left = join_with_spaces(
    element_ids
        .iter()
        .map(|element_id| get_render_function(*element_id))
        .flat_map(|render| render(context).0),
);
```

hamrik commented:
Okay, I have reversed the inclusion and also replaced the hand-rolled interspersing iterator with your helper function.

hamrik commented:
Debating whether to increase spacing between elements from one to two spaces.

hamrik commented:
Increased spacing to two spaces, which is actually looks closer to how spacing looked before the PR.

hamrik commented:
Double spacing breaks consistency next to `mode` when color-modes is on. Argh! 

This PR became a mess, so I'm closing it and will make a new one with a smaller scope.
